src/DotVVM.Framework/Compilation/CompiledAssemblyCache.cs

```diff
@@ -11,6 +24,79 @@ public class CompiledAssemblyCache
         private readonly ConcurrentDictionary<Assembly, MetadataReference> cachedAssemblyMetadata = new ConcurrentDictionary<Assembly, MetadataReference>();
         private readonly ConcurrentDictionary<string, Assembly> cachedAssemblies = new ConcurrentDictionary<string, Assembly>();
 
+        private readonly DotvvmConfiguration configuration;
+
+        public static CompiledAssemblyCache? Instance { get; private set; }
+
+        public CompiledAssemblyCache(DotvvmConfiguration configuration)
+        {
+            this.configuration = configuration;
+
+#if DotNetCore
+            System.Runtime.Loader.AssemblyLoadContext.Default.Resolving += DefaultOnResolving;
+#else
+            AppDomain.CurrentDomain.AssemblyResolve += TryResolveAssembly;
+#endif
+
+            foreach (var assembly in BuildReferencedAssembliesCache())
+            {
+                GetAssemblyMetadata(assembly);
+            }
+
+            if (Instance == null)
+            {
+                // normally, the constructor is not called multiple times as this service is singleton; only in unit tests, we create it multiple times
+                Instance = this;
+            }
+        }
+
+        private IEnumerable<Assembly> BuildReferencedAssembliesCache()
+        {
+            var diAssembly = typeof(ServiceCollection).Assembly;
+
+            var references = diAssembly.GetReferencedAssemblies().Select(Assembly.Load)
+                .Concat(configuration.Markup.Assemblies.Select(e => Assembly.Load(new AssemblyName(e))))
```

Comment:
Shall we load the referenced assemblies recursively?

Reply:
Not sure - the only related issue I am aware of was with calling the static constructors (the static constructors on base types are not called automatically) which I'll handle elsewhere. But I think we don't need to explicitly load referenced assemblies.